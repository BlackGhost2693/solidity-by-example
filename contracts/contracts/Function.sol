pragma solidity ^0.5.16;

contract Function {
    // cannot use map for inputs
    // function mapInput(mapping(uint => uint) memory map) public {
    // }

    // can use multi dimensional fixed size array for input
    function multiDimFixedSizeArrayInput(uint[9][9] memory _arr) public {
    }

    // cannot use multi dimensional dynamic size array as input
    // function multiDimDynamicSizeArrayInput(uint[][] memory _arr) public {
    // }

    // can use array for input
    function arrayInput(uint[] memory _arr) public {
    }

    mapping(uint => uint) map;
    uint[] arr;
    uint[9][9] arr2DFixed;
    uint[][] arr2D;

    // cannot return map for outputs
    // function mapOutput() public returns (mapping(uint => uint) memory) {
    //     return map;
    // }

    // can use multi dimensional fixed size array for output
    function multiDimFixedSizeArrayOutput() public view returns (uint[9][9] memory) {
        return arr2DFixed;
    }

    // cannot use multi dimensional dynamic size array as output
    // function multiDimDynamicSizeArrayOutput() public returns (uint[][] memory) {
    //     return arr2D;
    // }

    // can use array for output
    function arrayOutput() public view returns (uint[] memory) {
        return arr;
    }

    // Functions can return multiple values.
    function returnMultipleVals() public pure returns (uint, bool, uint) {
        return (1, true, 2);
    }

    // Return values can be named. This is useful when a contract
    // interacts with web3.js where you can access the returned
    // values by name instead of the returned order.
    function named() public pure returns (uint x, bool b, uint y) {
        return (1, true, 2);
    }

    // Return values can be assigned to their name.
    // In this case the return statement can be omitted;
    function assigned() public pure returns (uint x, bool b, uint y) {
        x = 1;
        b = true;
        y = 2;
    }

    // Use destructing assignment when calling another
    // function that returns multiple values.
    function destructingAssigments() public pure returns (uint, bool, uint, uint, uint) {
        (uint i, bool b, uint j) =  returnMultipleVals();

        // Values can be left out.
        (uint x, , uint y) = (4, 5, 6);

        return (i, b, j, x, y);
    }
}
