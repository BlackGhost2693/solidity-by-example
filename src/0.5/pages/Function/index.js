import React from "react"
import Example from "../../components/Example"
import html from "./index.html.js"

export default function(props) {
  return (
    <Example
      title="Function"
      description="Example of how to write functions in Solidity"
      html={html}
    />
  )
}
