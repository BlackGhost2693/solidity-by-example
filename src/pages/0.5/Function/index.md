There are several ways to return outputs from a function.

Public functions cannot accept certain data types as inputs or outputs

```solidity
{{{Function}}}
```
