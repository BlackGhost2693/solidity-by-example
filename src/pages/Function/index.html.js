const html = `<p>There are several ways to return outputs from a function.</p>
<p>Public functions cannot accept certain data types as inputs or outputs</p>
<pre><code class="language-solidity">pragma solidity ^0.5.16;

contract Function {
    // Functions can return multiple values.
    function returnMany() public pure returns (uint, bool, uint) {
        return (1, true, 2);
    }

    // Return values can be named.
    function named() public pure returns (uint x, bool b, uint y) {
        return (1, true, 2);
    }

    // Return values can be assigned to their name.
    // In this case the return statement can be omitted.
    function assigned() public pure returns (uint x, bool b, uint y) {
        x = 1;
        b = true;
        y = 2;
    }

    // Use destructing assignment when calling another
    // function that returns multiple values.
    function destructingAssigments()
        public pure returns (uint, bool, uint, uint, uint)
    {
        (uint i, bool b, uint j) =  returnMany();

        // Values can be left out.
        (uint x, , uint y) = (4, 5, 6);

        return (i, b, j, x, y);
    }

    // Cannot use map for neither input nor output

    // Can use array for input
    function arrayInput(uint[] memory _arr) public {
    }

    // Can use array for output
    uint[] public arr;

    function arrayOutput() public view returns (uint[] memory) {
        return arr;
    }
}
</code></pre>
`

export default html
