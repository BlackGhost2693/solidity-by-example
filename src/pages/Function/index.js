import React, { Component } from 'react'
import Example from '../../components/Example'
import html from './index.html.js'

class Function extends Component {
  render() {
    return (
      <Example
        title="Function"
        html={html}
      />
    )
  }
}

export default Function
