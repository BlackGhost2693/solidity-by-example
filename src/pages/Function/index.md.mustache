In Solidity, public functions cannot accept certain data types as inputs or outputs

```solidity
{{{Function}}}
```
