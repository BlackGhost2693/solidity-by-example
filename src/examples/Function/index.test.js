import React from "react"
import { shallow } from "enzyme"
import Function from "./index"

test("it renders", () => {
  const component = shallow(<Function />)

  expect(component).toMatchSnapshot()
})
