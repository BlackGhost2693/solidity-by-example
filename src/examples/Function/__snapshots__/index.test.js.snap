// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<Example
  html="<p>You can interact with a smart contract by calling the functions defined in the contract.</p>
<pre><code class=\\"language-solidity\\">pragma solidity ^0.5.1;

contract Function {
    uint counter;

    // This function costs Ether.
    function increment(uint i) public {
        counter += i;
    }

    // This function is free when it is called from an account.
    function add(uint i, uint j) public pure returns (uint) {
        return i + j;
    }

    // Functions can return multiple values.
    function returnMultipleVals() public pure returns (uint, uint) {
        return (1, 2);
    }

    // Return values can be named. This is useful when a contract
    // interacts with web3.js where you can access the returned
    // values by name instead of the returned order.
    function named() public pure returns (uint x, uint y) {
        return (1, 2);
    }

    // Return values can be assigned to their name.
    // In this case the return statement can be omitted;
    function assigned() public pure returns (uint x, uint y) {
        x = 1;
        y = 2;
    }

    // Use destructing assignment when calling another
    // function that returns multiple values.
    function destructingAssigments() public pure returns (uint) {
        (uint x, uint y) = (1, 2);
        // Values can be left out.
        (uint a, , uint b) = (4, 5, 6);
        (uint i,) =  returnMultipleVals();

        return x + y + a + b + i;
    }
}
</code></pre>
"
  title="Function"
/>
`;
